Feature/shape union

pablosebastiasaez opened the pull request:
Dear Regionmask-Team,

In continuation to the issue “percentage overlap #38” in your GitHub repository (https://github.com/regionmask/regionmask/issues/38), I have developed 3 functions that you might find of interest. 
These functions are suited to calculate areal weighted masks for regular grids, rotated pole grids and irregular grids. Additionally, I added a function to combine different shape files into one to finally calculate the weights for the whole region of interest. These functions are accessed via the method-selection. The methods are named as: “weights_default”, “weights_rot_pole” and “weights_irregular” (the latter one is still under construction).

The main function is called _mask_weights(lon,lat,shapefile_region) in mask.py. This function reads the lat/lon-Grid from a given NetCDF file and creates polygons that describe each grid cell in the region of interest defined by the lon, lat parameters of the function call. Then these gridded polygons are intersected with the provided shapefile and the area of the intersection is computed. Finally, the weighted area is returned as fractions between 0 and 1. The sum of all weights for all gridboxes equals to one.

The function _mask_weights is called by either _Default(lon,lat,polygons,numbers,fill=np.NaN) or _Rotated_Pole(lon,lat,polygons,numbers,fill=np.NaN) in mask.py. The difference is that _Default works directly with regular grids and _Rotated_Pole first projects the given shapefile in the rotated pole grid with a projection string. To proceed with this projection, the function asks the user for the projection string as extra input. The information on how to write the string is provided; the actual values of the rotated pole have to be known by the user/ provided by extracting them from the NetCDF. Finally, both call the function _mask_weights. 

If there is interest in running these functions for the case of multiple given shapefiles, one can provide a list of the shapefiles to be masked. This automatically merges all shapes in the function _prepare_gdf_for_mask (_geopandas.py) with the function “unary_union” imported from shapely.

Under regionmask/regionmask/examples I added a Jupyter Notebook with examples of the functions above. Under regionmask/regionmask/tests/test_geopandas.py I added my method under the test for “test_mask_geopandas”. 

To notice is the consideration of the behaviour at edges. My functions skips the edges only returning the region contained. Compared to rasterize or shapely, my method takes one step to the inside not only in the lower and left edge, but in the upper and right too.

If you are interested, I would like to contribute with these extensions to regionmask. Your module is extensively used in our institute GERICS. If you have an idea for further improvements or suggestions, please let us know. It would be great to contribute to this really nice and useful module!

Best regards,
Pablo Sebastia Saez and the GERICS team

mathause commented:
Dear @pablosebastiasaez,
thank you very much for your interest and the contribution. I fear that we will not be able to use your contribution as is (it's too big, not actually tested, duplicates code, some parts are not easy to read, different API from what we would choose) and would require a very considerable amount of work from our and your side to make fit to include in regionmask. It demonstrates some interesting ideas, though, that are worth investigating.

Intersecting grid-box polygons does return exact overlaps, but in my experience it is prohibitively slow. This is a main motivation behind the approximate approach outlined in #38 which is super fast. How long does it take to create a mask with this method, e.g. for the srex regions and a 2°x2° regular grid? One potential way to speed this up is to only do the exact calculation for grid cells that are on the border of the regions, see https://gist.github.com/perrygeo/721040f8545272832a42

best, Mathias

pablosebastiasaez commented:
Dear Regionmask-Team,
Thank you very much for your answer and feedback. Indeed, this still is a work in progress, we are still completing the tests and improving our functions, therefore any feedback is welcome. Our goal is to contribute with our functions to regionmask, which we are 
using extensively. We would like to enable regionmask users to create 
fractional masks by using the same module and similar interface/function 
calls.
Regarding the efficiency of the method, we computed the test you suggested for the srex region of the coast of south america. The function shared with you needs around 0.72 seconds. Nonetheless, considering the comment of only masking the region of interest speeds up the computation and now the function needs 0.03 seconds. Thank you for pointing this out!
Please let us know if you have any further suggestions to improve our functions, we are open and willing to make the needed changes. 
Best regards,
Pablo Sebastia and the GERICS team

pablosebastiasaez commented:
@mathause  we have updated and simplified our functions. Now they are faster and take  additional details into account, like the coordinate transformation with the function wrap_lon. Also we want to stress the feature of creating masks on rotated pole grids with the implemented method "weights_rot_pole". 
Please let us know if you have interest in our functions and or have further suggestions.
Bests, Pablo/ GERICS 
Now, we are working on increasing the test coverage.
